access/discord/config.go

```diff
@@ -0,0 +1,129 @@
+	"net/url"
+	"strings"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/pelletier/go-toml"
+
+	"github.com/gravitational/teleport-plugins/access/common"
+	"github.com/gravitational/teleport-plugins/lib"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+type DiscordConfig struct {
+	common.BaseConfig
+	Discord common.GenericAPIConfig
+}
+
+// LoadDiscordConfig reads the config file, initializes a new DiscordConfig struct object, and returns it.
+// Optionally returns an error if the file is not readable, or if file format is invalid.
+func LoadDiscordConfig(filepath string) (*DiscordConfig, error) {
+	t, err := toml.LoadFile(filepath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conf := &DiscordConfig{}
+	if err := t.Unmarshal(conf); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if strings.HasPrefix(conf.Discord.Token, "/") {
+		conf.Discord.Token, err = lib.ReadPassword(conf.Discord.Token)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	if err := conf.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return conf, nil
+}
+
+// CheckAndSetDefaults checks the config struct for any logical errors, and sets default values
+// if some values are missing.
+// If critical values are missing and we can't set defaults for them — this will return an error.
+func (c *DiscordConfig) CheckAndSetDefaults() error {
+	if err := c.Teleport.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if c.Discord.Token == "" {
+		return trace.BadParameter("missing required value discord.token")
+	}
+	if c.Log.Output == "" {
+		c.Log.Output = "stderr"
+	}
+	if c.Log.Severity == "" {
+		c.Log.Severity = "info"
+	}
+
+	if len(c.Recipients) == 0 {
+		return trace.BadParameter("missing required value role_to_recipients.")
+	} else if len(c.Recipients[types.Wildcard]) == 0 {
+		return trace.BadParameter("missing required value role_to_recipients[%v].", types.Wildcard)
+	}
+
+	return nil
+}
+
+// NewBot initializes the new Discord message generator (DiscordBot)
+func (c *DiscordConfig) NewBot(clusterName, webProxyAddr string) (common.MessagingBot, error) {
+	var (
+		webProxyURL *url.URL
+		err         error
+	)
+	if webProxyAddr != "" {
+		if webProxyURL, err = lib.AddrToURL(webProxyAddr); err != nil {
+			return DiscordBot{}, trace.Wrap(err)
+		}
+	}
+
+	token := "Bot " + c.Discord.Token
+
+	client := resty.
+		NewWithClient(&http.Client{
+			Timeout: discordHTTPTimeout,
+			Transport: &http.Transport{
+				MaxConnsPerHost:     discordMaxConns,
+				MaxIdleConnsPerHost: discordMaxConns,
+			},
+		}).
+		SetHeader("Content-Type", "application/json").
+		SetHeader("Accept", "application/json").
+		SetHeader("Authorization", token)
+
+	// APIURL parameter is set only in tests
+	if endpoint := c.Discord.APIURL; endpoint != "" {
+		client.SetHostURL(endpoint)
+	} else {
+		client.SetHostURL("https://discord.com/api/")
+		client.OnAfterResponse(onAfterResponseDiscord)
```

Comment:
Why don't we set this handler when running tests?

Reply:
It's taken from the slack access plugin, I think it'sbecause the mockserver does not set `code` when returning errors, which would cause the unmarshalling to `DiscordResponse` to fail.